### DISCUSSION TITLE
[next.config.js] environment passes into the client code

### DISCUSSION THREAD

**Discussion by akomm** [vercel/next.js#24741] [Help]
### What version of Next.js are you using?

10.1.3

### What version of Node.js are you using?

14.15.0

### What browser are you using?

firefox

### What operating system are you using?

linux/ubuntu 18 LTS

### How are you deploying your application?

next start

### Describe the Bug

Environment variables provided via `next.config.js` are not removed from client code. If that is not a bug and was made by the assumption of no secrets defined in the versioned file, I think its a false assumption, because you can dynamically load not versioned secrets inside the config file using for example `fs`.

### Expected Behavior

Only env variables starting with `NEXT_PUBLIC_` should be substituted into the client code.

### To Reproduce

See codesandbox: https://codesandbox.io/s/intelligent-butterfly-w1j92?file=/pages/index.js

**Comment by jwugit** [comment]
are you mixing 
https://nextjs.org/docs/basic-features/environment-variables

with 
https://nextjs.org/docs/api-reference/next.config.js/environment-variables

when you actually want this
https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration

the docs says
```
To add environment variables to the JavaScript bundle, open next.config.js and add the env config:
```

which is what you seem to be doing

**Comment by mattcarlotta** [comment]
What @jwugit states is correct. 

Ideally, you should use `.env` files to load [environment specific variables](https://nextjs.org/docs/basic-features/environment-variables#loading-environment-variables), which makes it clearer that Envs with [NEXT_PUBLIC](https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser) are available to both client and server and added to the bundle; whereas Envs without `NEXT_PUBLIC` are only available to the server.

In the documentation, [this way of loading Envs](https://nextjs.org/docs/api-reference/next.config.js/environment-variables) specifically states

> To add environment variables **_to the JavaScript bundle_**, open next.config.js and add [an] env config
> ```js
> module.exports = {
>  env: {
>    customKey: 'my-value',
>  },
>}
>```

In other words, these Envs aren't required to use `NEXT_PUBLIC` to be accessible by the client and server; therefore, they should be consider **global** and **publicly** available. 

The above way of loading Envs pre-dates using [@next/env](https://github.com/vercel/next.js/tree/canary/packages/next-env) to load Envs with the `NEXT_PUBLIC` rule. In the past, if you wanted to load Envs and expose them to the browser, you had to do it manually:

```js
// load .env with Envs
require("dotenv").config({
  path:
    process.env.NODE_ENV === "development"
      ? ".env.development"
      : ".env.production",
});
const { DefinePlugin } = require("webpack");


module.exports = {
  // apply Envs through the "env" property...
  env: {
    EXAMPLE: process.env.EXAMPLE,
  },
 // ...or by utilizing the "DefinePlugin" plugin
  webpack(config, { isServer }) {

    if (!isServer)
      config.plugins.push(
        new DefinePlugin({
          "process.env": {
             EXAMPLE: process.env.EXAMPLE,
           },
        }),
      );

    return config;
  },
};
```
My guess as to why it's still in the documentation is most likely for backwards compatibility for older projects. That said, it's a legacy approach which technically should be deprecated in favor of loading Envs through `.env` -- since it no longer requires a custom `next.config.js` file (unless you need to use a runtime Env that shouldn't be added to the bundle).

Example of utilizing Envs through a `.env` file, the `env` config and `publicRunTimeConfig`:
[![Edit Next Envs Example](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/next-envs-example-rucu8?fontsize=14&hidenavigation=1&theme=dark)

**Reply by akomm** [reply]
Thanks. I went through the whole docs, so I am aware of the 3 options. My intuition was circling around runtime config, but I was wandering why the env was there and did not want mix configuration over env and runtime, so decided to use env for all. But my inner intuition actually told me, that runtime config for providing the project root path sounds more fitting for the purpose... Sometimes need to listen to the intuition. Thanks for clearing up :+1:
